### ISSUE TITLE
[css-overflow] Which kind of boxes accept max-lines?

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#2408]
In b39a87c4b31cb5feab06ad60dce5cbae4f530048 @fantasai changed [`max-lines`](https://drafts.csswg.org/css-overflow-3/#max-lines) from applying to "all non-inline elements" to "block containers (excluding multi-column containers)".

But what about things like flex or grid containers? I'm not convinced that this feature should work across formatting context boundaries, but the spec mentions flex or grid containers, so it seems it does

> line boxes are counted in document order OR visual order (grid order in grid containers, flex order in flex containers, and document order in all other display types)

So if my understanding is correct, it seems I can use

```html
<div style="max-lines: 1">
  <p style="display: flex; width: 0">Lorem ipsum</p>
</div>
```
and I will get

> Lorem

But if I use

```html
<div style="max-lines: 1; display: flex; width: 0">Lorem ipsum</div>
```

then `max-lines` does not apply and I will get

> Lorem
> ipsum

Doesn't make much sense to me.

**Comment by frivoal** [comment]
@Loirooriol Thanks for the lightning fast review. You caught @fantasai &amp; I mid brainstorming :) We continued working on this a few hours later, the spec has changed a bit since you reviewed it.

> I'm not convinced that this feature should work across formatting context boundaries

We agreed, so

> the spec mentions flex or grid containers

is no longer true, and the spec now says:

> The contents of descendants that are scroll containers or that are not block containers are skipped over while counting line boxes. 

Based on your comments, I believe you agree, So I'm closing this issue as accepted. If I misunderstood you, please reopen.

**Comment by Loirooriol** [comment]
OK, now it makes more sense! But it says

> The contents of descendants that are scroll containers or that are not block containers are skipped over while counting line boxes. 

Should block containers descendants that are not scroll containers but establish a new BFC be skipped too?

**Comment by frivoal** [comment]
@fantasai and I considered that possibility, but we thought it was better not to skip:
* We did not think it would make a big difference in implementation complexity
* Counting lines in nested BFCs doesn't seem to introduce strange corner cases
* The various reasons block container get BFCfied did not seem to be particularly relevant to the issue of counting lines, do we expect authors will prefer it this way
* Authors can always turn their BFCs into some other kind of formatting contexts if they want to exclude them (e.g. a single item vertical flexbox)

For scroll containers, we were more on the edge. Many of the same arguments work to some extend, but seemed less convincing. For these, I could easily go either way, depending on use cases.

**Comment by Loirooriol** [comment]
> Counting lines in nested BFCs doesn't seem to introduce strange corner cases

Really? I would skip lines at least from out-of-flow and inline-block nested BFCs. Otherwise please explain more precisely how they are treated, because I don't think it's clear at all.

**Comment by frivoal** [comment]
> I would skip lines at least from out-of-flow

I agree, and the spec already says to only count in-flow line boxes.

> and inline-block nested BFCs.

Aha! didn't think of that one, good catch. I agree should be skipped as well.

**Comment by Loirooriol** [comment]
> the spec already says to only count in-flow line boxes.

OK! But not sure if it's completely clear. What if you have a float (out-of-flow) which contains an in-flow block container with line boxes?

I think you should say that line boxes are considered to be out-of-flow if there is some ancestor of the line box that is a descendant of the element with `max-lines` and which is out-of-flow.

**Comment by Loirooriol** [comment]
I filed #2429 to discuss which descendants should be skipped, so probably this can be closed.

**Comment by MatsPalmgren** [comment]
Are there any technical reasons for why Grid and Flexbox containers are excluded?  I don't really see any since both supports being fragmented. We could count grid rows and flex lines as "lines".

**Comment by jonjohnjohnson** [comment]
I'm guessing grid and flexbox are excluded for reasons of their "lines" being structurally/semantically different than "lines" of text needing to be cut off, most likely in the middle of the node, as well as providing an eliding visual. Just seems to be a different use case, unless children/blocks of flexbox/grid need to be "cut" in half visually and I can't even imagine how eliding that would be represented?

**Comment by frivoal** [comment]
Right. The idea for not going through Flexbox/Grid/Table formatting context was a combination of being more complicated, and not seeming that it is particularly related to the original use case. It is still possible to apply this to flex items or grid items, just not to the container as a whole.

We could make it work if we wanted to, but it doesn't seem worth the effort.

**Comment by frivoal** [comment]
Ok, so since #2429 is open to talk about descendants, this is back to only talking about which type of boxes it applies to, and based on the reasoning above, I would like to close with no (further) change (i.e. accept the initial comment from @Loirooriol which has already been baked into the spec, and reject @MatsPalmgren suggestion that we make it apply to different sorts of lines, such as grid tracks).
